### PR TITLE
check local store when executing get_providers

### PR DESCRIPTION
**Pull request by stuartnelson3**
closes https://github.com/libp2p/rust-libp2p/issues/2209

remaining TODO on removing expired providers when checking the local store

### CODE REVIEW

File: protocols/kad/src/behaviour.rs [hunk 1]
```diff
@@ -912,9 +912,17 @@ where
...
     /// reported via [`KademliaEvent::OutboundQueryCompleted{QueryResult::GetProviders}`].
     pub fn get_providers(&mut self, key: record::Key) -> QueryId {
+        let providers = self
+            .store
+            .providers(&key)
+            .into_iter()
+            .filter(|p| !p.is_expired(Instant::now()))
+            .map(|p| p.provider)
+            .collect();
+        // TODO(stn): call remove_provider on expired providers?
```
Comment (mxinden) [inline protocols/kad/src/behaviour.rs:922]:
```suggestion
```

I just double checked, expired provider records are periodically garbage collected (see below). Thus calling `remove_provider` here is not needed.

https://github.com/libp2p/rust-libp2p/blob/e83e1b3d0be822f58d705bc724c87f267c806150/protocols/kad/src/jobs.rs#L308-L314

File: protocols/kad/CHANGELOG.md [hunk 2]
```diff
@@ -8,7 +8,10 @@
 - Introduce `KademliaStoreInserts` option, which allows to filter records (see
   [PR 2163]).
 
+- Check local store first when `get_providers` (see [PR 2221]).
```
Comment (mxinden) [inline protocols/kad/CHANGELOG.md:11]:
```suggestion
- Check local store first when `Kademlia::get_providers` (see [PR 2221]).
```

File: protocols/kad/src/behaviour.rs [hunk 3]
```diff
@@ -912,9 +912,17 @@ where
...
         let info = QueryInfo::GetProviders {
             key: key.clone(),
-            providers: HashSet::new(),
+            providers: providers,
```
Comment (mxinden) [inline protocols/kad/src/behaviour.rs:925]:
```suggestion
            providers,
```

File: protocols/kad/src/behaviour/test.rs [hunk 4]
```diff
@@ -1317,3 +1317,52 @@ fn network_behaviour_inject_address_change() {
         kademlia.addresses_of_peer(&remote_peer_id),
     );
 }
+
+#[test]
+fn get_providers() {
+    fn prop(key: record::Key) {
+        let config = KademliaConfig::default();
+        let (_, mut single_swarm) = build_node_with_config(config);
```
Comment (mxinden) [inline protocols/kad/src/behaviour/test.rs:1325]:
```suggestion
        let (_, mut single_swarm) = build_node();
```

**Review by mxinden — COMMENTED** [review]
Couple of comments. Looks good to me overall. Thanks @stuartnelson3 for your first contributions. Hope for many more to come :angel:

**Comment by stuartnelson3** [comment]
Thanks for the review and the help!

**Review by mxinden — APPROVED** [review]

**Pull request merged**